packages/lodestar/src/network/encoder.ts

```diff
@@ -0,0 +1,131 @@
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {ERR_INVALID_REQ, MAX_CHUNK_SIZE, Method, ReqRespEncoding} from "../constants";
+import {RequestBody, ResponseBody} from "@chainsafe/lodestar-types";
+import {IReqRespEncoder} from "./encoders/interface";
+import {getRequestMethodSSZType, getResponseMethodSSZType} from "./util";
+import {SszEncoder} from "./encoders/ssz";
+import {SnappyEncoder} from "./encoders/snappy";
+import * as varint from "varint";
+import {Response, ResponseChunk} from "./interface";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import BufferList from "bl";
+
+export class ReqRespEncoder {
+
+  public readonly encoding: ReqRespEncoding;
+
+  private readonly config: IBeaconConfig;
+
+  constructor(config: IBeaconConfig, encoding: ReqRespEncoding) {
+    this.config = config;
+    this.encoding = encoding;
+  }
+
+  public encodeRequest(method: Method, data: RequestBody): Buffer {
+    const type = getRequestMethodSSZType(this.config, method);
+    const encoders = this.getEncoders(method, this.encoding);
+    const encodedPayload =  encoders.reduce((result: unknown, encoder) => {
+      return encoder.encode(type, result);
+    }, data) as Buffer;
+    return this.writeLengthPrefixed(encodedPayload);
+  }
+  
+  public decodeRequest(method: Method, data: Buffer): RequestBody {
+    data = this.readLengthPrefixed(data);
+    const type = getRequestMethodSSZType(this.config, method);
+    //decoding is done backwards
+    const encoders = this.getEncoders(method, this.encoding).reverse();
+    return encoders.reduce((result: unknown, encoder) => {
+      return encoder.decode(type, result);
+    }, data) as RequestBody;
+  }
+
+  public encodeResponse(
+    method: Method
+  ): ((source: AsyncIterable<ResponseChunk|Partial<Response>>) => AsyncGenerator<Buffer>) {
```

Comment:
Why are we iterating over `Partial<Response>`? I thought responses are always `Arrayish<ResponseChunk>`?

Reply:
When we handle request we are yielding each ResponseChunk, but sometimes, we yield error or empty Response.
If we hook up stream from database we could stream blocks directly to response